applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -0,0 +1,856 @@
+    const Variable<double>& rVariable, std::vector<double>& rValues, const ProcessInfo& rCurrentProcessInfo)
+{
+    if (rValues.size() != 1)
+        rValues.resize(1);
+
+    if (rVariable == PRESSURE_COEFFICIENT)
+    {
+        rValues[0] = PotentialFlowUtilities::ComputePerturbationCompressiblePressureCoefficient<Dim, NumNodes>(*this, rCurrentProcessInfo);
+    }
+    else if (rVariable == DENSITY)
+    {
+        rValues[0] = ComputeDensity(rCurrentProcessInfo);
+    }
+    else if (rVariable == MACH)
+    {
+        rValues[0] = PotentialFlowUtilities::ComputePerturbationLocalMachNumber<Dim, NumNodes>(*this, rCurrentProcessInfo);
+    }
+    else if (rVariable == SOUND_VELOCITY)
+    {
+        rValues[0] = PotentialFlowUtilities::ComputePerturbationLocalSpeedOfSound<Dim, NumNodes>(*this, rCurrentProcessInfo);
+    }
+    else if (rVariable == WAKE)
+    {
+        const TransonicPerturbationPotentialFlowElement& r_this = *this;
+        rValues[0] = r_this.GetValue(WAKE);
+    }
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::GetValueOnIntegrationPoints(
+    const Variable<int>& rVariable, std::vector<int>& rValues, const ProcessInfo& rCurrentProcessInfo)
+{
+    if (rValues.size() != 1)
+        rValues.resize(1);
+    if (rVariable == TRAILING_EDGE)
+        rValues[0] = this->GetValue(TRAILING_EDGE);
+    else if (rVariable == KUTTA)
+        rValues[0] = this->GetValue(KUTTA);
+    else if (rVariable == WAKE)
+        rValues[0] = this->GetValue(WAKE);
+    else if (rVariable == ZERO_VELOCITY_CONDITION)
+        rValues[0] = this->GetValue(ZERO_VELOCITY_CONDITION);
+    else if (rVariable == TRAILING_EDGE_ELEMENT)
+        rValues[0] = this->GetValue(TRAILING_EDGE_ELEMENT);
+    else if (rVariable == DECOUPLED_TRAILING_EDGE_ELEMENT)
+        rValues[0] = this->GetValue(DECOUPLED_TRAILING_EDGE_ELEMENT);
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::GetValueOnIntegrationPoints(
+    const Variable<array_1d<double, 3>>& rVariable,
+    std::vector<array_1d<double, 3>>& rValues,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    if (rValues.size() != 1)
+        rValues.resize(1);
+    if (rVariable == VELOCITY){
+        const array_1d<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
+        array_1d<double, 3> v(3, 0.0);
+        array_1d<double, Dim> vaux = PotentialFlowUtilities::ComputeVelocity<Dim, NumNodes>(*this);
+        for (unsigned int k = 0; k < Dim; k++)
+            v[k] = vaux[k] + free_stream_velocity[k];
+        rValues[0] = v;
+    }
+    else if (rVariable == PERTURBATION_VELOCITY)
+    {
+        array_1d<double, 3> v(3, 0.0);
+        array_1d<double, Dim> vaux = PotentialFlowUtilities::ComputeVelocity<Dim,NumNodes>(*this);
+        for (unsigned int k = 0; k < Dim; k++)
+            v[k] = vaux[k];
+        rValues[0] = v;
+    }
+}
```

Comment:
As far as i can understand this functions are completely (or almost completely) copied and pasted from other elements in the application.

The point is that this element derives from the base `Element` in the core. I think it should derive from one base element in you application. If the base elements already there can't be the parent class, an intermediate layer should be created.

Reply:
or a utility where you put those functions

Reply:
@rubenzorrilla thanks for the review and the helpful comments!



> The point is that this element derives from the base `Element` in the core. I think it should derive from one base element in you application. If the base elements already there can't be the parent class, an intermediate layer should be created.

You are definitely right in here. We need a major restructuring of the elements inheritance in the PotentialFlowApp. Also most of the elements implement three different types of element in one element and they should be separated. I would suggest doing these changes in a different PR since they are not part of @eloisabaez master thesis.

Reply:
OK then. In any case I'd put this as a priority. The more elements you have in there the more complex the transition would be (some years ago we had a rally tough experience with the FluidDynamicsApplication).